File: src/api/bundles.rs
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::Json;
use bytes::Bytes;

use super::error::{ApiError, bad_request, not_found, server_error};
use crate::app::AppState;
use crate::backend::Backend;
use crate::bundle;

#[derive(serde::Serialize)]
pub struct UploadResponse {
    pub bundle_id: String,
    pub task_id: String,
}

pub async fn upload_bundle<B: Backend>(
    State(state): State<AppState<B>>,
    Path(app_id): Path<String>,
    body: Bytes,
) -> Result<(StatusCode, Json<UploadResponse>), ApiError> {
    // 1. Validate app exists
    let app = crate::db::sqlite::resolve_app(&state.db, &app_id)
        .await
        .map_err(|e| server_error(format!("db error: {e}")))?
        .ok_or_else(|| not_found(format!("app {app_id} not found")))?;

    // 2. Validate body is not empty
    if body.is_empty() {
        return Err(bad_request("empty bundle body".into()));
    }

    // Note: bundle size is enforced at the router level via
    // axum::extract::DefaultBodyLimit — requests exceeding
    // max_bundle_size are rejected with 413 before reaching
    // this handler. See api_router() setup.

    // 3. Generate IDs
    let bundle_id = uuid::Uuid::new_v4().to_string();
    let task_id = uuid::Uuid::new_v4().to_string();

    // 4. Write archive atomically
    let base = &state.config.storage.bundle_server_path;
    let paths = bundle::write_archive(base, &app.id, &bundle_id, body)
        .await
        .map_err(|e| server_error(format!("write archive: {e}")))?;

    // 5. Unpack
    if let Err(e) = bundle::unpack_archive(&paths).await {
        bundle::delete_bundle_files(&paths).await;
        return Err(server_error(format!("unpack: {e}")));
    }

    // 6. Create library dir
    if let Err(e) = bundle::create_library_dir(&paths).await {
        bundle::delete_bundle_files(&paths).await;
        return Err(server_error(format!("create lib dir: {e}")));
    }

    // 7. Insert bundle row (status = pending)
    if let Err(e) = crate::db::sqlite::create_bundle(
        &state.db,
        &bundle_id,
        &app.id,
        paths.archive.to_str().unwrap(),
    )
    .await
    {
        bundle::delete_bundle_files(&paths).await;
        return Err(server_error(format!("create bundle row: {e}")));
    }

    // 8. Create task in TaskStore
    let task_sender = state.task_store.create(task_id.clone());

    // 9. Get image name
    let image = state
        .config
        .docker
        .as_ref()
        .map(|d| d.image.clone())
        .unwrap_or_else(|| crate::config::DEFAULT_IMAGE.into());

    // 10. Spawn async restore
    bundle::restore::spawn_restore(bundle::restore::RestoreParams {
        backend: state.backend.clone(),
        pool: state.db.clone(),
        task_store: state.task_store.clone(),
        task_sender,
        app_id: app.id,
        bundle_id: bundle_id.clone(),
        paths,
        image,
        retention: state.config.storage.bundle_retention,
        bundle_server_path: state.config.storage.bundle_server_path.clone(),
    });

    // 11. Return 202
    Ok((
        StatusCode::ACCEPTED,
        Json(UploadResponse { bundle_id, task_id }),
    ))
}

pub async fn list_bundles<B: Backend>(
    State(state): State<AppState<B>>,
    Path(app_id): Path<String>,
) -> Result<Json<Vec<crate::db::sqlite::BundleRow>>, ApiError> {
    let app = crate::db::sqlite::resolve_app(&state.db, &app_id)
        .await
        .map_err(|e| server_error(format!("db error: {e}")))?
        .ok_or_else(|| not_found(format!("app {app_id} not found")))?;

    let bundles = crate::db::sqlite::list_bundles_by_app(&state.db, &app.id)
        .await
        .map_err(|e| server_error(format!("db error: {e}")))?;
    Ok(Json(bundles))
}

File: internal/api/bundles.go
package api

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"
	"path/filepath"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/cynkra/blockyard/internal/audit"
	"github.com/cynkra/blockyard/internal/auth"
	"github.com/cynkra/blockyard/internal/bundle"
	"github.com/cynkra/blockyard/internal/manifest"
	"github.com/cynkra/blockyard/internal/server"
	"github.com/cynkra/blockyard/internal/telemetry"
)

// UploadBundle uploads a new bundle archive for an app.
//
//	@Summary		Upload bundle
//	@Description	Upload a tar.gz bundle archive. Triggers async restore (package installation). Track progress via the returned task_id.
//	@Tags			bundles
//	@Accept			application/gzip
//	@Produce		json
//	@Param			id	path		string	true	"App ID (UUID) or name"
//	@Success		202	{object}	uploadBundleResponse
//	@Failure		400	{object}	errorResponse
//	@Failure		404	{object}	errorResponse
//	@Failure		413	{object}	errorResponse
//	@Failure		500	{object}	errorResponse
//	@Security		BearerAuth
//	@Router			/apps/{id}/bundles [post]
func UploadBundle(srv *server.Server) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		caller := auth.CallerFromContext(r.Context())
		appID := chi.URLParam(r, "id")

		app, relation, ok := resolveAppRelation(srv, w, caller, appID)
		if !ok {
			return
		}

		if !relation.CanDeploy() {
			notFound(w, "app not found")
			return
		}

		// Enforce body size limit
		r.Body = http.MaxBytesReader(w, r.Body, srv.Config.Storage.MaxBundleSize)

		// Generate IDs
		bundleID := uuid.New().String()
		taskID := uuid.New().String()
		slog.Info("bundle upload started", "app_id", appID, "bundle_id", bundleID) //nolint:gosec // G706: slog structured logging handles this

		// Stream archive to disk
		paths := bundle.NewBundlePaths(srv.Config.Storage.BundleServerPath, app.ID, bundleID)
		if err := bundle.WriteArchive(paths, r.Body); err != nil {
			var maxBytesErr *http.MaxBytesError
			if errors.As(err, &maxBytesErr) {
				writeError(w, http.StatusRequestEntityTooLarge, "payload_too_large",
					"bundle exceeds max_bundle_size")
				return
			}
			serverError(w, "write archive: "+err.Error())
			return
		}

		// Unpack
		if err := bundle.UnpackArchive(paths); err != nil {
			bundle.DeleteFiles(paths)
			serverError(w, "unpack: "+err.Error())
			return
		}

		// Validate entrypoint
		if err := bundle.ValidateEntrypoint(paths); err != nil {
			bundle.DeleteFiles(paths)
			badRequest(w, err.Error())
			return
		}

		// Create library dir
		if err := bundle.CreateLibraryDir(paths); err != nil {
			bundle.DeleteFiles(paths)
			serverError(w, "create library dir: "+err.Error())
			return
		}

		// Insert bundle row (status = pending)
		// Determine if bundle has pinned dependencies by checking the manifest.
		bundlePinned := false
		manifestPath := filepath.Join(paths.Unpacked, "manifest.json")
		if m, mErr := manifest.Read(manifestPath); mErr == nil {
			bundlePinned = m.IsPinned()
		}
		deployedBy := ""
		if caller != nil {
			deployedBy = caller.Sub
		}
		if _, err := srv.DB.CreateBundle(bundleID, app.ID, deployedBy, bundlePinned); err != nil {
			bundle.DeleteFiles(paths)
			serverError(w, "create bundle row: "+err.Error())
			return
		}

		// Create task in TaskStore
		sender := srv.Tasks.Create(taskID, app.ID)

		// Spawn async restore
		actorSub := "anonymous"
		if caller != nil {
			actorSub = caller.Sub
		}
		bundle.SpawnRestore(bundle.RestoreParams{
			Backend:          srv.Backend,
			DB:               srv.DB,
			Tasks:            srv.Tasks,
			Sender:           sender,
			AppID:            app.ID,
			BundleID:         bundleID,
			Paths:            paths,
			Image:            server.AppImage(app, srv.Config.Docker.Image),
			PakVersion:       srv.Config.Docker.PakVersion,
			PakCachePath:     filepath.Join(srv.Config.Storage.BundleServerPath, ".pak-cache"),
			BuilderVersion:   srv.Version,
			BuilderCachePath: filepath.Join(srv.Config.Storage.BundleServerPath, ".by-builder-cache"),
			Retention:        srv.Config.Storage.BundleRetention,
			BasePath:         srv.Config.Storage.BundleServerPath,
			Store:            srv.PkgStore,
			AuditLog:         srv.AuditLog,
			AuditActor:       actorSub,
			WG:               srv.RestoreWG,
		})

		telemetry.BundlesUploaded.Inc()
		if srv.AuditLog != nil {
			srv.AuditLog.Emit(auditEntry(r, audit.ActionBundleUpload, app.ID,
				map[string]any{"bundle_id": bundleID}))
		}

		slog.Info("bundle upload accepted, restore spawned", //nolint:gosec // G706: slog structured logging handles this
			"app_id", appID, "bundle_id", bundleID, "task_id", taskID)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusAccepted)
		json.NewEncoder(w).Encode(map[string]string{
			"bundle_id": bundleID,
			"task_id":   taskID,
		})
	}
}

// ListBundles lists all bundles for an app.
//
//	@Summary		List bundles
//	@Description	List all bundles for an app, ordered by upload time.
//	@Tags			bundles
//	@Produce		json
//	@Param			id	path		string	true	"App ID (UUID) or name"
//	@Success		200	{object}	bundleListResponse
//	@Failure		404	{object}	errorResponse
//	@Failure		500	{object}	errorResponse
//	@Security		BearerAuth
//	@Router			/apps/{id}/bundles [get]
func ListBundles(srv *server.Server) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		caller := auth.CallerFromContext(r.Context())
		appID := chi.URLParam(r, "id")

		app, relation, ok := resolveAppRelation(srv, w, caller, appID)
		if !ok {
			return
		}

		// RBAC: collaborator+ required (tightened from "any access").
		if !relation.CanDeploy() {
			notFound(w, "app not found")
			return
		}

		bundles, err := srv.DB.ListBundlesByApp(app.ID)
		if err != nil {
			serverError(w, "db error: "+err.Error())
			return
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{"bundles": bundles})
	}
}
